### DISCUSSION TITLE
Resume a lua thread from rust and make it error inside lua

### DISCUSSION THREAD

**Discussion by filiptibell** [mlua-rs/mlua#252] [Q&A] [answered]
Hi,

I'm trying to implement a task scheduler to mix and match blocking lua code with async rust code without doing busy polling on the lua side, but I can't figure out how to resume a lua thread while making it error inside of lua. I need to do this to let lua scripts handle errors that may happen in Rust code using `pcall`, `xpcall` etc.

Example code from the scheduler:
https://github.com/filiptibell/lune/blob/b1e08bf813ece005be5f05b3f8d9d88e52768d12/packages/lib/src/lua/task/scheduler.rs#L244

A relevant test script is here and currently fails:
https://github.com/filiptibell/lune/blob/d2ff0783a5f81c8ecbd38536163a6605c8764d05/tests/net/serve/requests.luau#L31

**Comment by khvzak** [comment, marked as answer]
I'm not sure that I fully understood the problem, do you have a smaller example?

In general if you throw a error in lua coroutine, it unwinds the stack and there is no function to jump back:
```lua
co = coroutine.create(function ()
    coroutine.yield("1")
    error("bla")
    coroutine.yield("2")
end)

-- prints: true 1
print(coroutine.resume(co))
-- prints: false	test.lua:3: bla
print(coroutine.resume(co))
-- prints: false	cannot resume dead coroutine
print(coroutine.resume(co))
```

**Reply by filiptibell** [reply]
Basically I have a thread that is ready to be resumed, and an external lua result produced from somewhere else in rust. I would like to resume the thread and have it throw an error even though in the lua code there is nowhere that produces an error. Something like this:

```lua
local thread = coroutine.create(function()
    coroutine.yield(1)
    coroutine.yield(2)
end)

return thread
```

```rust
let thread: LuaThread = ...
thread.resume(()) // resumes at 1
thread.resume_error(LuaError::external()) // throws an error at 2
```

**Reply by khvzak** [reply]
Probably the easiest way is set a hook (interrupt for Luau) just before resuming and throw error inside it. It should look like it produced from coroutine.
